### ISSUE TITLE
Reengineer the tree APIs to be more interfacey

### ISSUE THREAD

**Issue by bwateratmsft** [microsoft/vscode-azuretools#1127]
Dependent on #1126. It's a longer-term stretch goal but we should try to make the tree APIs and such use the interfaces internally, so that `AzExtTreeDataProvider`, etc. are no longer tightly coupled to the implementations.

If we can do this cleanly without exploding the interfaces, we might be able to avoid needing a "v3" tree API.

**Comment by alexweininger** [comment]
I made [changed `AzExtParentTreeItem` and `AzExtTreeDataProvider`](https://github.com/microsoft/vscode-azuretools/compare/main...alex/no-cache) so that the result of `getChildren` is no longer cached.

I made sure to use the updated utils package in both RGs and SWA for my testing.

From my exploration of RGs + SWA, the tree felt and behaved exactly the same aside from the focus/unfocus tree item actions. As I'll explain below, I'm afraid we might need to keep at least a watered down version of our caching logic to maintain the current fast and smooth focus/unfocus experience.

Video comparing focusing a group with and without caching
<table>
<tr>
	<td>Caching enabled (current experience)
	<td>Caching disabled
<tr>
	<td>

https://user-images.githubusercontent.com/12476526/174913831-52fa5fe5-12c4-4f1c-ab27-4aa08f08ddad.mov
	
</td>
	<td>

https://user-images.githubusercontent.com/12476526/174913480-cd1b10b0-f29c-47e6-bec9-d878515969fe.mov

</table>

Here's what I think is happening:
1. In order to focus/unfocus a group, we must fire the `onDidChangeTreeDataEmitter` event.
2. VS Code makes a series of calls to `getChildren` for the tree items that are expanded.
3. Since we're no longer caching, calls to `getChildren` are async and delay the UI update enough to degrade the experience.

<details>
<summary>Debug logs</summary>


For my own sanity I logged when `getChildren` and `refreshUIOnly` were called:

```
Refreshing UI only root 
Getting children of root
Getting children of aweininger dev sub
Getting children of Static Web Apps
Getting children of angular-basic-api
Getting children of Production
Getting children of Application Settings
Getting children of Functions
Getting children of root
Getting children of aweininger dev sub
Getting children of Static Webo r Apps
Getting children of angular-basic-api
Getting children of Production
Getting children of Application Settings
Getting children of Functions
```

The first log is when [we call `refreshUIOnly` upon focusing the group](https://github.com/microsoft/vscode-azureresourcegroups/blob/5d589179ab2fa142fa74e80470a8861de54c7d12/src/tree/SubscriptionTreeItem.ts#L217)

</details>

I think the snappiness of the focus/unfocus actions is important warrant keeping our own cache. I couldn't think of any good workarounds to this problem, since we don't have much control over when VS Code makes calls to `getChildren`.
